### PR TITLE
Support emb/reranker/seq_cls padding_free

### PR DESCRIPTION
**Pull request by tastelikefeet**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

Write the detail information belongs to this PR.

## Experiment results

Paste your experiment result here(if needed).

### CODE REVIEW

File: swift/utils/np_utils.py [hunk 1]
```diff
@@ -28,6 +28,8 @@ def get_seed(random_state: Optional[np.random.RandomState] = None) -> int:
 
 def stat_array(array: Union[np.ndarray, List[int], 'torch.Tensor']) -> Tuple[Dict[str, float], str]:
     if isinstance(array, list):
+        if isinstance(array[0], list):
```
Comment (Jintao-Huang) [inline swift/utils/np_utils.py]: array可能是空list

File: swift/trainers/trainer_factory.py [hunk 2]
```diff
@@ -67,4 +67,8 @@ def get_training_args(cls, args):
                 args_dict.pop(k)
 
         args._prepare_training_args(args_dict)
-        return training_args_cls(**args_dict)
+        training_args = training_args_cls(**args_dict)
+        training_args.padding_side = args.padding_side
```
Comment (Jintao-Huang) [inline swift/trainers/trainer_factory.py]: 可以把参数加载 SwiftArgumentsMixin中

File: swift/trainers/mixin.py [hunk 3]
```diff
@@ -584,6 +586,92 @@ def clip_grad_norm_(self, parameters, *args, **kwargs):
...
+                    return transformers_seq_cls_forward(model, origin_forward=inner_forward, **kwargs)
+
+                model.forward = MethodType(seq_cls_forward, model)
+            elif self.args.task_type == 'reranker':
+                llm_model = get_llm_model(self.model, model_meta=self.model.model_meta)
+
+                def revert_padding_free_hook(module, args, input, output):
+                    return revert_padding_free(output, input, self.args.padding_side)
+
+                llm_model.register_forward_hook(revert_padding_free_hook, with_kwargs=True, prepend=True)
```
Comment (Jintao-Huang) [inline swift/trainers/mixin.py:666]: 感觉这里或许可以 统一一下

File: swift/llm/model/patcher.py [hunk 4]
```diff
@@ -167,6 +171,80 @@ def get_lm_head_model(model, model_meta=None, lm_heads=None):
...
+        batch_size = input_ids.shape[0]
+    else:
+        batch_size = inputs_embeds.shape[0]
+
+    if self.config.pad_token_id is None and batch_size != 1:
+        raise ValueError('Cannot handle batch sizes > 1 if no padding token is defined.')
+    if self.config.pad_token_id is None:
+        sequence_lengths = -1
+    else:
+        if output.get('attention_mask') is not None:
```
Comment (Jintao-Huang) [inline swift/llm/model/patcher.py:197]: 这里方便加个注释嘛

File: examples/train/reranker/train_generative_reranker.sh
Comment (0russwest0) [inline examples/train/reranker/train_generative_reranker.sh:1]: 为什么要加padding_side，是默认的padding_side传不进mixin吗？

File: examples/train/reranker/train_reranker.sh
Comment (0russwest0) [inline examples/train/reranker/train_reranker.sh:1]: 传统reranker padding左右都可以

Comment (tastelikefeet) [reply examples/train/reranker/train_reranker.sh:1]: 确实，我改下

File: examples/train/reranker/train_generative_reranker.sh
Comment (tastelikefeet) [reply examples/train/reranker/train_generative_reranker.sh:1]: 作为一个显式的提醒，加了一个comment：
# only support --padding_side left

File: swift/utils/np_utils.py (hunk 1 above)
Comment (tastelikefeet) [reply swift/utils/np_utils.py]: 已修复

File: swift/trainers/mixin.py (hunk 3 above)
Comment (tastelikefeet) [reply swift/trainers/mixin.py:666]: 这里我觉得保持现状比较好，因为后面感觉可能对每个任务有特有代码，到时候还要拆出来

**Comment by ExtReMLapin** [comment]
So i gave a try with seq_cls 

Vram usage is much lower, which is expected, but training also much slower,I can increase batch size from 24 to 256, but after two pochs, for example the accuracy is at 0.08 (loss=4.0) while without padding_free we reach an accuracy of 0.22 at epoch 1.0 (loss = 2.7)

**Comment by tastelikefeet** [comment]
> So i gave a try with seq_cls 
> 
> 
> 
> Vram usage is much lower, which is expected, but training also much slower,I can increase batch size from 24 to 256, but after two pochs, for example the accuracy is at 0.08 (loss=4.0) while without padding_free we reach an accuracy of 0.22 at epoch 1.0 (loss = 2.7)

Can you share me your command line please? And your model and dataset format

**Comment by ExtReMLapin** [comment]
Since my message I discovered it was very likely linked to the learning rate that was not adapted to the new params.

I increased the batch size without lowering the learning rate, so yes out of the box, increasing batch size x10 slows down convergence, but dividing lr by 10 fixes the issue.
